adapters/audienceNetwork/facebook.go

```diff
@@ -37,267 +41,416 @@ func (a *FacebookAdapter) Name() string {
+			Message: fmt.Sprintf("Unexpected status code %d with error message '%s'", response.StatusCode, msg),
+		}}
 	}
 
-	requests := make([]bytes.Buffer, len(ortbRequests))
-	for i, ortbRequest := range ortbRequests {
-		e = json.NewEncoder(&requests[i]).Encode(ortbRequest)
-		if e != nil {
-			return nil, e
-		}
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
 	}
 
-	ch := make(chan adapters.CallOneResult)
-
-	for i := range requests {
-		go func(bidder *pbs.PBSBidder, reqJSON bytes.Buffer) {
-			result, err := a.callOne(ctx, reqJSON)
-			result.Error = err
-			if result.Bid != nil {
-				result.Bid.BidderCode = bidder.BidderCode
-				unit := bidder.LookupAdUnit(result.Bid.AdUnitCode)
-				if unit != nil {
-					result.Bid.Width = unit.Sizes[0].W
-					result.Bid.Height = unit.Sizes[0].H
-				}
-				result.Bid.BidID = bidder.LookupBidID(result.Bid.AdUnitCode)
-				if result.Bid.BidID == "" {
-					result.Error = &errortypes.BadServerResponse{
-						Message: fmt.Sprintf("Unknown ad unit code '%s'", result.Bid.AdUnitCode),
-					}
-					result.Bid = nil
-				}
+	out := adapters.NewBidderResponseWithBidsCapacity(4)
+	var errs []error
+
+	for _, seatbid := range bidResp.SeatBid {
+		for _, bid := range seatbid.Bid {
+			if bid.AdM == "" {
+				errs = append(errs, &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Bid %s missing 'adm'", bid.ID),
+				})
+				continue
 			}
-			ch <- result
-		}(bidder, requests[i])
-	}
 
-	var err error
+			var obj facebookAdMarkup
+			if err := json.Unmarshal([]byte(bid.AdM), &obj); err != nil {
+				errs = append(errs, &errortypes.BadServerResponse{
+					Message: err.Error(),
+				})
+				continue
+			}
 
-	bids := make(pbs.PBSBidSlice, 0)
-	for i := 0; i < len(requests); i++ {
-		result := <-ch
-		if result.Bid != nil {
-			bids = append(bids, result.Bid)
-		}
-		if req.IsDebug {
-			debug := &pbs.BidderDebug{
-				RequestURI:   a.URI,
-				RequestBody:  requests[i].String(),
-				StatusCode:   result.StatusCode,
-				ResponseBody: result.ResponseBody,
+			if obj.BidId == "" {
+				errs = append(errs, &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("bid %s missing 'bid_id' in 'adm'", bid.ID),
+				})
+				continue
 			}
-			bidder.Debug = append(bidder.Debug, debug)
+
+			bid.AdID = obj.BidId
+			bid.CrID = obj.BidId
+
+			out.Bids = append(out.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: resolveBidType(&bid, request),
+			})
 		}
-		if result.Error != nil {
-			err = result.Error
+	}
+
+	return out, errs
+}
+
+func resolveBidType(bid *openrtb.Bid, req *openrtb.BidRequest) openrtb_ext.BidType {
+	for _, imp := range req.Imp {
+		if bid.ImpID == imp.ID {
+			if typ, ok := resolveImpType(&imp); ok {
+				return typ
+			}
+
+			panic("processing an invalid impression")
 		}
 	}
 
-	if len(bids) == 0 {
-		return nil, err
+	panic(fmt.Sprintf("invalid bid imp ID %s", bid.ImpID))
```

Comment:
Please add a bit more description to why the impression is invalid. A future dev debugging this thanks you in advance. :)

Reply:
done